### ISSUE TITLE
Custom UITableViewCell with dynamic number of buttons 

### ISSUE THREAD

**Issue by engmsaleh** [PureLayout/PureLayout#89]
I have created a custom UITableViewCell which display dynamic number of buttons in horizontal access using the following initWithStyle
```
- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier
{
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        emailBtn = [UIButton newAutoLayoutView];
        [emailBtn setImage:[UIImage imageNamed:@"emailBtn"] forState:UIControlStateNormal];
        [emailBtn setImage:[UIImage imageNamed:@"emailBtn"] forState:UIControlStateSelected];
        [emailBtn addTarget:self action:@selector(emailBtnPressed:) forControlEvents:UIControlEventTouchUpInside];
        
        facebookBtn = [UIButton newAutoLayoutView];
        [facebookBtn setImage:[UIImage imageNamed:@"fbBtn"] forState:UIControlStateNormal];
        [facebookBtn setImage:[UIImage imageNamed:@"fbBtn"] forState:UIControlStateSelected];
        [facebookBtn addTarget:self action:@selector(facebookBtnPressed:) forControlEvents:UIControlEventTouchUpInside];
        
        twitterBtn = [UIButton newAutoLayoutView];
        [twitterBtn setImage:[UIImage imageNamed:@"twBtn"] forState:UIControlStateNormal];
        [twitterBtn setImage:[UIImage imageNamed:@"twBtn"] forState:UIControlStateSelected];
        [twitterBtn addTarget:self action:@selector(twitterBtnPressed:) forControlEvents:UIControlEventTouchUpInside];
        
        linkedinBtn = [UIButton newAutoLayoutView];
        [linkedinBtn setImage:[UIImage imageNamed:@"linkedInBtn"] forState:UIControlStateNormal];
        [linkedinBtn setImage:[UIImage imageNamed:@"linkedInBtn"] forState:UIControlStateSelected];
        [linkedinBtn addTarget:self action:@selector(linkedinBtnPressed:) forControlEvents:UIControlEventTouchUpInside];
        
        self.contentView.backgroundColor = [UIColor orangeColor];
        
        [self.contentView addSubview:emailBtn];
        [self.contentView addSubview:facebookBtn];
        [self.contentView addSubview:twitterBtn];
        [self.contentView addSubview:linkedinBtn];

        
        self.btnsArray = @[emailBtn,facebookBtn,twitterBtn,linkedinBtn];
        [self setNeedsUpdateConstraints]; // bootstrap Auto Layout
    }
    
    return self;
}
```

It works well and it display 4 buttons aligned in horizontal access , the problem I'm facing is that I need to make this buttons dynamic depending in data retrieved from API , for example if there are no Facebook link , Cell should display only 3 buttons ....
The problem that  
```
        [self setNeedsUpdateConstraints]; // bootstrap Auto Layout
```
is called in initializing the cell before I could configure cell data to determine the true number of buttons and their links.

Could any have suggestion for current situation?

**Comment by mickeyreiss** [comment]
It seems like there a number of ways to achieve this layout. Without a deeper understanding of the context, it's difficult to make a recommendation.  For example, you might even choose to use `UICollectionView` or `UIStackView`. It depends on the level of complexity, how precisely you need to control the layout, etc.

If you post a concrete solution, including PureLayout code snippets, along with the problems you experience with this solution, I think we'll be able to help you more.

**Comment by engmsaleh** [comment]
@mickeyreiss Please check attached screenshoot , this is basically a user profile screen. The Row of discussion is the Row that include social buttons , I'm using PureLayout to align these 4 buttons along horizontal cell axis as shown in the screenshoot.
![example](https://cloud.githubusercontent.com/assets/462150/9705666/09f83b2a-54d5-11e5-84a7-897749decbcf.png)
  
Here is the code used for making this cell layout
```
- (void)updateConstraints
{
    [super updateConstraints];
    if (!self.didSetupConstraints) {
        horizontalLayoutConstraints = [UIView autoCreateConstraintsWithoutInstalling:^{
            [self.btnsArray autoSetViewsDimension:ALDimensionHeight toSize:45.0];
            [self.btnsArray autoDistributeViewsAlongAxis:ALAxisHorizontal
                                               alignedTo:ALAttributeHorizontal
                                        withFixedSpacing:30.0
                                            insetSpacing:YES
                                            matchedSizes:YES];
            [[self.btnsArray firstObject] autoAlignAxisToSuperviewAxis:ALAxisHorizontal];
        }];
        
        [horizontalLayoutConstraints autoInstallConstraints];
        
        [[self.btnsArray firstObject] autoPinEdgeToSuperviewEdge:ALEdgeTop withInset:10];
        [[self.btnsArray firstObject] autoPinEdgeToSuperviewEdge:ALEdgeBottom withInset:10];

        self.didSetupConstraints = YES;
    }
}
```
My problem is that I want to make self.btnsArray to be filled dynamic according to which buttons is to be available , currently I have 
```
```

**Comment by smileyborg** [comment]
If I'm understanding this correctly, you have an asynchronous operation (network load) that happens before you have the data available to determine the actual UI state and layout of your cell. This is something that can be solved many ways. Generally speaking, I would recommend using a dedicated "loading" UI (perhaps a separate view that is shown before the real content is ready). Then once your data is available, you access the visible cells in the table view, and update the state of each cell so that it is configured with the real content, in addition to hiding your "loading" view.

As far as auto layout and constraints as long as the cell height isn't changing, things are pretty easy: you can just call `cell setNeedsUpdateConstraints` which will end up running your custom cell's `updateConstraints` method again, at which point you should remove any constraints no longer needed for the "loading" view/etc, and set up new constraints based on the actual data and content displayed in the cell.

So to summarize, there's no way to avoid the initial call to `updateConstraints` when the cell is first displayed even if you don't have the data for it yet...you need to just do some configuration, which you then replace with the real content once it's available.

Does that make sense?

**Comment by engmsaleh** [comment]
Thanks @smileyborg  Yup, it make sense , I will try to make that solution and update the issue with my result.

**Comment by smileyborg** [comment]
Closing this for now, please reopen if you still need more help with this.
